### DISCUSSION TITLE
[General] L2 Interoperability - zkSync Hyperchains vs Polygon AggLayer

### DISCUSSION THREAD

**Discussion by goodacres** [zkSync-Community-Hub/zksync-developers#421] [ZK Stack] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

Hey, apologies if this is not the best place to ask this question - if there is somewhere more appropriate please let me know for next time.

I have been delving deeper into ZK Rollups and am curious about interoperability between different ZKRs such as zkSync, Polygon, Scroll etc. 

I think I understand the concept of L2 & L3 hyperchains relying on the same shared L1 bridge for finality (Daniel Lumi had a great talk at EthDenver here - https://www.youtube.com/watch?v=zLj-m7A3Fxs). 

I am slightly confused however about when L2 ZKR projects talk about interoperability (hyperchains & hyperbridging in the case of zkSync). Will the various L2 ZKR projects utilise the same shared L1 bridge? Or are the benefits of L2 interoperability isolated only to each individual rollup? 

Mr. Lumi also tweeted recently (https://twitter.com/zkLumi/status/1770052604266996193) about the importance of building on a shared bridge like the ZK Stack hyperchain. I am not 100% sure but this would lead me to believe that different L2 projects will not be inherently interoperable unless traditional bridging with token locking & minting is setup.

Any guidance or further reading materials would be much appreciated!

Thanks in advance :)

### What is the specific issue or error you're encountering?

NA

### Can you share the error messages or logs you're receiving, if any?

NA

### Have you made any recent changes to the contract before encountering this issue?

NA

### Are there any external libraries or contracts that your contract interacts with?

NA

### Can you provide the relevant portions of your contract code where the issue is occurring?

NA

### Have you tried to isolate the problem, and if so, what were the results?

NA

### What steps have you already taken to try to resolve the issue?

NA

**Comment by zk-Lumi** [comment, marked as answer]
Hi @goodacres,

Happy to hear that you liked the speech 😊

To have safe, cheap, and trustless interoperability, you need to use ZK validity proofs, and have all the chains need to be deployed on the same shared bridge contract. 

This is unlikely to happen across different ZK rollups as the trust assumptions between all different L2s are very different (note, they all currently have their own separate bridge contracts for this reason).

This is the reason we are building the zkSync hyperchain using [ZK Stack](https://zkstack.io/). This allows all chains connected to the hyperchain shared bridge to have seamless and trustless interoperability between each other, effectively creating one big internet of value.

You can learn more at a high level about our hyperchain & hyperbridging [here](https://docs.zksync.io/zk-stack/concepts/hyperchains-hyperscaling.html). Much more content will be coming out about our implementation in the coming months.

For the longer term, Ethereum foundation is also researching a concept called [universal synchronous composability](https://www.youtube.com/watch?v=MnsjUZo7RRI), which would make synchronous interoperability between different ZK L2s (zkSync, Scroll, Polygon) possible. However even with universal synchronous composability, interoperability in the hyperchain will be much cheaper (shared bridge + hypermessages are part of state commitments) and a smoother experience. Additionally with hyperchains, operators won't need to run full nodes for every single chain in the shared bridge, unlike with 'USC' where the builder role (PBS) will need to run full nodes for every single separate L2 + as well as every single 'hyperchain' in their respective shared bridges. Hyperchain operators also won't need any precommitment from connected chains.

Hope this helped!

PS: Appreciate your feedback, I will make a note to add more slides on our shared bridge and how it is important, in future versions of the speech 👍

**Reply by goodacres** [reply]
Amazing! Thanks so much for getting back & apologies for the delay in reply. This really helps clear things up for me :)
